### PR TITLE
add `naked_asm!` macro for use in `#[naked]` functions

### PR DESCRIPTION
**Pull request by folkertdev**
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/128651)*

tracking issue: https://github.com/rust-lang/rust/issues/90957

Adds the `core::arch::naked_asm` macro, to be used in `#[naked]` functions, but providing better error messages and a place to explain the restrictions on assembly in naked functions.

This PR does not yet require that the `naked_asm!` macro is used inside of `#[naked]` functions:

- the `asm!` macro can still be used in `#[naked]` functions currently, with the same restrictions and error messages as before.
- the `naked_asm!` macro can be used outside of `#[naked]` functions. It has not yet been decided whether that should be allowed long-term.

In this PR, the parsing code of `naked_asm!` now enforces the restrictions on assembly in naked functions, with the exception of checking that the `noreturn` option is specified. It also has not currently been decided if `noreturn` should be implicit or not.

This PR looks large because it touches a bunch of tests. The code changes are mostly straightforward I think: we now have 3 flavors of assembly macro, and that information must be propagated through the parsing code and error messages.

cc @Lokathor  

r? @Amanieu

### CODE REVIEW

**Comment by Lokathor** [comment]
My own vote is that the noreturn option *should not* be placed by users into the naked_asm macro. The same option should not mean one thing for asm and then nearly the opposite in naked_asm. That's needlessly confusing.

There are two reasonable designs that spring to mind:
* naked_asm uses an alternative token such as nakedreturn, which still lets the reader know that there's some funny control flow going on, but which can be documented separately and compiler checked separately.
* naked_asm simply doesn't have users write any return related option word at all.

**Comment by folkertdev** [comment]
I agree, and think we can get away with making the `noreturn` implicit. Naked functions are an advanced feature anyway, so if you're writing `naked_asm!` it's up to you to make sure the safety conditions are satisfied.

But also, I don't want to touch the validation logic in this PR, it can be handled when we disallow the use of `asm!` in naked functions.

**Review by Amanieu — APPROVED** [review]
LGTM, but I would make one change and remove the `noreturn` attribute from `naked_asm`, changing it to always return `!`.

The reason for this is that we want existing users of naked function to transition to using `naked_asm` instead and we want to eliminate `noreturn` as part of this transition.

r=me

**Comment by folkertdev** [comment]
cool, we discussed `noreturn` a bit in the comments here earlier, and it looks like everyone is in agreement that the `noreturn` should be dropped.

I did not implement that here yet because then `naked_asm!` will fail the validation pass that is currently applied to `asm!` which mandates that the `noreturn` is present. So to make `naked_asm!` work without `noreturn`, that validation pass needs to be removed and in practice that means that `asm!` should also be disallowed in naked functions. That was always the plan of course, I just tried to break it up into smaller parts.

But, I'll line up those changes and then we can see how to merge them: either in 2 parts or as one combined unit.

**Comment by Amanieu** [comment]
It's probably fine to go ahead and disallow `asm` in naked functions as part of this PR. It will be easier this way.

File: compiler/rustc_ast/src/ast.rs
Comment (folkertdev) [inline compiler/rustc_ast/src/ast.rs:2275]: this means the parser will now disallow the `noreturn` option

File: compiler/rustc_hir_typeck/src/expr.rs
Comment (folkertdev) [inline compiler/rustc_hir_typeck/src/expr.rs:3314]: from a type system perspective, `naked_asm!`  always diverges and is inferred as `!`, just like a standard `asm!` with the `options(noreturn)` option.

File: compiler/rustc_passes/src/errors.rs
Comment (folkertdev) [inline compiler/rustc_passes/src/errors.rs:1210]: this diff is weird, but the old message is no longer needed, and we do now need a nice error message for when an `asm!` is used where `naked_asm!` should be used.

File: compiler/rustc_passes/src/naked_functions.rs
Comment (folkertdev) [inline compiler/rustc_passes/src/naked_functions.rs:140]: I know there is some concern over raw `BytePos` manipulation (because utf8), but I think this is fine?

File: compiler/rustc_passes/src/naked_functions.rs
Comment (folkertdev) [inline compiler/rustc_passes/src/naked_functions.rs:218]: this check is actually not needed any more: the parser will error out in all cases (invalid operands and invalid options). Should we keep it, or just remove it?

Comment (Amanieu) [reply compiler/rustc_passes/src/naked_functions.rs:218]: Yes this should just be removed.

File: compiler/rustc_passes/src/naked_functions.rs
Comment (Amanieu) [reply compiler/rustc_passes/src/naked_functions.rs:140]: The problem here is that the lint isn't fully machine-applicable because:
- naked_asm may not have been imported
- asm may have been imported with a different name: `use core::arch::asm as renamed_asm`

Comment (folkertdev) [reply compiler/rustc_passes/src/naked_functions.rs:140]: hmm, right. So there is no way to actually get the name of the macro here (besides looking at the source). Should I just remove the span and maybe adjust the error message to point to `naked_asm!`?

Comment (folkertdev) [reply compiler/rustc_passes/src/naked_functions.rs]: I removed the explicit suggestion (the error message still suggests `naked_asm!` of course)

**Comment by folkertdev** [comment]
well that pinged a bunch of people.

The thing I had to solve was that the `InlineAsmOptions::NORETURN` value is (ab)used to answer the question "will this asm block diverge?", which is used in various places in type checking and codegen. The `naked_asm!` macro does not set this option, but does diverge. In hindsight that option maybe should have been named `diverges`? anyway, the current solution is a bit noisy, so if anyone has better ideas let me know.

File: compiler/rustc_codegen_cranelift/src/base.rs
Comment (bjorn3) [inline compiler/rustc_codegen_cranelift/src/base.rs:517]: This should never be a naked function. We already handle naked functions specially above.

File: compiler/rustc_codegen_cranelift/src/base.rs
Comment (bjorn3) [inline compiler/rustc_codegen_cranelift/src/base.rs:60]: Maybe make this check for NakedAsm, such that anything else will cause the unreachable!() below to be hit?

File: compiler/rustc_codegen_cranelift/src/base.rs
Comment (folkertdev) [reply compiler/rustc_codegen_cranelift/src/base.rs:517]: you're right, but at the moment `naked_asm!` can occur outside of naked functions.

There isn't really a technical reason to disallow `naked_asm!` anywhere `asm!` is allowed, but there also isn't really much practical use because `asm!` could just be used instead. @Amanieu and others: thoughts?

**Comment by folkertdev** [comment]
also, how do I fix this

```
error[E0787]: the `asm!` macro is not allowed in naked functions
##[error]  --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/compiler_builtins-0.1.123/src/x86_64.rs:17:9
   |
17 | /         core::arch::asm!(
18 | |             "push   %rcx",
19 | |             "push   %rax",
20 | |             "cmp    $0x1000,%rax",
...  |
35 | |             options(noreturn, att_syntax)
36 | |         );
   | |_________^ consider using the `naked_asm!` macro instead
```

kind of a bootstrapping problem

File: compiler/rustc_codegen_cranelift/src/base.rs
Comment (Amanieu) [reply compiler/rustc_codegen_cranelift/src/base.rs]: `naked_asm` should be specifically disallowed anywhere except in naked functions. It doesn't make sense in other contexts.

**Comment by Amanieu** [comment]
> also, how do I fix this
> 
> ```
> error[E0787]: the `asm!` macro is not allowed in naked functions
> ##[error]  --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/compiler_builtins-0.1.123/src/x86_64.rs:17:9
>    |
> 17 | /         core::arch::asm!(
> 18 | |             "push   %rcx",
> 19 | |             "push   %rax",
> 20 | |             "cmp    $0x1000,%rax",
> ...  |
> 35 | |             options(noreturn, att_syntax)
> 36 | |         );
>    | |_________^ consider using the `naked_asm!` macro instead
> ```
> 
> kind of a bootstrapping problem

The best way would be to open a separate PR which just makes `naked_asm` a re-export of `asm`. Then we can update compiler-builtins to use `naked_asm`. Then we can land this PR along with an update to compiler-builtins.

**Comment by folkertdev** [comment]
this is blocked on compiler-builtins being updated now, the use of naked-asm was merged https://github.com/rust-lang/compiler-builtins/pull/686 but I guess it'll take a little while for that to trickle through.

**Comment by Amanieu** [comment]
[You can update compiler](https://github.com/rust-lang/compiler-builtins/pull/687) will publish a new release of compiler-builtins. Once that's merged, you can update compiler-builtins in this PR.

**Comment by folkertdev** [comment]
it looks like there is still some sort of bootstrapping issue

```
    Checking rustc-std-workspace-core v1.99.0 (/checkout/library/rustc-std-workspace-core)
error[E0433]: failed to resolve: could not find `naked_asm` in `arch`
   --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/compiler_builtins-0.1.126/src/aarch64_linux.rs:139:38
    |
130 | / macro_rules! compare_and_swap {
131 | |     ($ordering:ident, $bytes:tt, $name:ident) => {
132 | |         intrinsics! {
133 | |             #[maybe_use_optimized_c_shim]
...   |
139 | |                 unsafe { core::arch::naked_asm! {
    | |                                      ^^^^^^^^^ could not find `naked_asm` in `arch`
156 | |     };
157 | | }
    | |_- in this expansion of `compare_and_swap!` (#2)
```

This is using the new compiler-builtins version `0.1.126`, and `naked_asm!` has been in nightly for a bit now, so I'm not sure where the problem is. Could it be this `cfg(not(bootstrap))`? I tried removing that but it ran into other issues.

```
#[cfg(not(bootstrap))]
pub macro naked_asm("assembly template", $(operands,)* $(options($(option),*))?) {
    /* compiler built-in */
}
```

**Comment by bjorn3** [comment]
Can you add a `cfg(bootstrap)` `naked_asm!` macro which forwards to `asm!` and adds the `noreturn` option?

File: library/core/src/arch.rs
Comment (bjorn3) [inline library/core/src/arch.rs:34]: Try something like

```suggestion
    ($(arguments:tt)*) => { core::arch::asm!($(arguments)*, options(noreturn)) }
```

This will forward all tokens from the input and add `, options(noreturn)` as suffix.

**Comment by bjorn3** [comment]
> Would it be possible to get the current naked_asm! "into" bootstrap?

It will be on the next bootstrap bump. Current nightly branches to beta on October 11th, a couple of days after that the bootstrap bump should happen. The alternative is to request a beta backport of the PR that introduced `naked_asm!`.

**Comment by folkertdev** [comment]
hmm,  @Amanieu thoughts? I guess the options are

- a beta backport; is a beta backport for this sort of thing justified? My impression has been that it's usually just for bug fixes
- we could make the trailing commas in compiler-builtins consistent (so all or none have trailing commas), then the `macro_rules!` approach could work.
- we wait ~ a month

**Comment by folkertdev** [comment]
finally fixed all the CI issues. 

I think the only open question from my side is whether the `diverges` solution introduced in https://github.com/rust-lang/rust/pull/128651/commits/9eef0bf02f23079c635255d5d38d653b681a821b is worth it versus always setting the `noreturn` option behind the scenes.

File: library/core/src/arch.rs
Comment (tgross35) [inline library/core/src/arch.rs:33]: This means we will have two `naked_asm`s, one from this rule and the existing `pub macro naked_asm`. Is that intentional?

Comment (tgross35) [reply library/core/src/arch.rs:33]: Nevermind, I see the bootstrap config now.

**Comment by folkertdev** [comment]
the bootstrap `naked_asm!` is a standalone thing, and could be split out if that's useful

**Comment by tgross35** [comment]
Would you be up for putting that and the builtins update (still only to 0.1.126, can bump further separately) into a new PR?

**Comment by folkertdev** [comment]
yes, I'll do that tomorrow then

File: compiler/rustc_passes/src/naked_functions.rs
Comment (Amanieu) [inline compiler/rustc_passes/src/naked_functions.rs:219]: Make this a `span_bug!`.

**Review by Amanieu — APPROVED** [review]
LGTM with one minor nit.

r=me

**Pull request merged**
